### ISSUE TITLE
just a question

### ISSUE THREAD

**Issue by cpu0x00** [Yair-Men/ExplorerMen#1]
hello author,  first of all Excellent work :ok_hand: 

my question is, before checking wether or not watchdog mode is enabled, in Program.cs lines 123-149 you kill the process with the handle to it retrieved by its PID from `NtQuerySystemInformation()`, as follows:

```cs
            var lstHandles = GetHandles(PID // <--- this function DOES NOT return handle to process using the DeviceIoControl call
            if (lstHandles.Count == 0)
            {
                PrettyPrint("[!] No handles found", Info);
                CloseHandle(hDriver);
                return;
            }

#if DEBUG
            Console.WriteLine("=== Listing all handles ===");
            foreach (var handle in lstHandles)
            {
                var output = string.Format("UniqueProcessId: {0} , ObjectType: {1}, Handle Flags: {2}, Handle Value: 0x{3:x},  ObjectPointer: 0x{4:x}, Access Mask: 0x{5:x}",
                    handle.UniqueProcessId, handle.ObjectType, handle.HandleFlags, handle.HandleValue, handle.ObjectPointer, handle.AccessMask);
            }
            Console.WriteLine("[!] Total number of lstHandles: {0}", lstHandles.Count);
#endif



            foreach (var handle in lstHandles)
            {
                if (GetHandleType((IntPtr)(handle.HandleValue)) != null)
                {
                    Driver.CloseHandle(hDriver, (IntPtr)(handle.HandleValue), PID, (ulong)handle.ObjectPointer);
                }
            }
```

now after the watch dog check, you actually check for and use the kernel-land handle, lines 183-185

```cs
// this calls DeviceIoCtl with IOCTL_OPEN_PROTECTED_PROCESS_HANDLE
IntPtr hProc = userMode ? OpenProcess(PROCESS_SUSPEND_RESUME, false, (ulong)PID) : Driver.GetProcessHandle(hDriver, PID); 
NtSuspendProcess(hProc);
CloseHandle(hProc); 
```
what confused is that in the first example:
- a kernel-land process handle was not used 
- a kernel land CloseHandle was called to close said handle

however in the second example 
- a kernel land handle was optained 
- a normal CloseHandle was used to close it + why NtSuspendProcess was called ??

pardon me if thats a stupid question :sweat_smile: , its really confusing especially because im going through a C implementation and they seem a bit different

**Comment by Yair-Men** [comment]
Thanks for the feedback..
In general, there are some fixing I need to do, but can't find time for it now (For example, the code doesn't take into account cases where there can be multiple EDR processes - servers for example where multiple users connect and there is a process per logon)

As for your question, I hope  I can clarify the logic with a few words.

I did a research in the past about handles, and found out that an admin user can get process handle to any process (even PP/L)
If the access rights are one of three below:
1. PROCESS_QUERY_LIMITED_INFORMATION
2. PROCESS_SUSPEND_RESUME
3. Sorry can't remember the 3th :(

However, when suspending an AV/EDR process the entire OS freezes slowly until it stucks (my guess is that all processes hooked and since the EDR process is suspended, hooks can't be treated).

With that, if you suspend the EDR process as its spawns, it doesn't have enough time to send hooks so you can supsend it without freezing the OS.

The first time I kill the handles, I just collect the addresses of the handles (not asking for handle, just query its address) and passes them to the driver because that specific IOCTL (IRP) closes handle by a given handle's address.

In watch dog, I don't need to close all the process handles because I suspend it (Remember I can get process handle with access right PROCESS_SUSPEND_RESUME without any driver ?) and it eventually dies..

If for some reason, I couldn't get process handle from user mode then Im asking from the driver ti give me one..

So to summarize, Ill answer your list:
- I never asked (nor had to) kernel handle, just to find its address and pass it to the driver to close it
- Same here, the driver only need the address of the handle to close it (which can be acquired using `NtQuerySystemInformtion` from user-mode)
- A kernel handle obtained just in cases where user mode handle can't be obtained
- Watch dog is a long-living process, I'm closing handles to free spaces (This is my handle, not the handles that kills the EDR), about the `NtSuspendProcess` I've answered it above

Hope it clarifies a few thinks 🙂

**Comment by cpu0x00** [comment]
thank you for detailed response, feel free to close the issue
